### PR TITLE
llvm-r2.eclass: New eclass to unmess pkg_setup and make cross work finally

### PR DESCRIPTION
**Pull request by mgorny**
Just some initial bits. They work great for CMake, but Meson is crap, so it needs me to implement a lot of crap in `llvm-config`, even though we never really need it because we use dylib, so I don't know if I'm going to pursue this after all.

### CODE REVIEW

**Comment by mgorny** [comment]
CC @gentoo/llvm

File: eclass/llvm-r2.eclass [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+		die "LLVM ${LLVM_SLOT} not found installed in BROOT (expected: ${broot_prefix}/bin)"
+
+	llvm_fix_clang_version CC CPP CXX
+	# keep in sync with profiles/features/llvm/make.defaults!
+	llvm_fix_tool_path ADDR2LINE AR AS LD NM OBJCOPY OBJDUMP RANLIB
+	llvm_fix_tool_path READELF STRINGS STRIP
+
+	# Set LLVM_CONFIG to help Meson (bug #907965) but only do it
+	# for empty ESYSROOT (as a proxy for "are we cross-compiling?").
+	if [[ -z ${ESYSROOT} ]] ; then
```
[the remark this answers is not in the document]
Comment (mgorny) [reply eclass/llvm-r2.eclass:363]: Oh wait, I was supposed to remove that! It shouldn't be necessary, I think — or if anything, it should be replaced with our llvm-config script.

File: eclass/llvm-r2.eclass [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+llvm_chost_setup() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	local esysroot_prefix=$(get_llvm_prefix -d)
+	einfo "Using ${esysroot_prefix} for CHOST LLVM ${LLVM_SLOT}"
+	[[ -d ${esysroot_prefix} ]] ||
+		die "LLVM ${LLVM_SLOT} not found installed in ESYSROOT (expected: ${esysroot_prefix})"
+
+	# satisfies find_package() in CMake
+	export CMAKE_PREFIX_PATH="${sysroot_prefix}${CMAKE_PREFIX_PATH+:${CMAKE_PREFIX_PATH}}"
```
[the remark this answers is not in the document]
Comment (mgorny) [reply eclass/llvm-r2.eclass:390]: It's just a variable name. Wanted to keep it shorter.

File: eclass/llvm-utils.eclass [hunk 3]
```diff
@@ -113,17 +113,38 @@ llvm_fix_tool_path() {
...
+		-*)
+			die "${FUNCNAME}: invalid option: ${1}"
+			;;
+	esac
+
+	[[ ${#} -ne 1 ]] && die "Usage: ${FUNCNAME} [-b|-d] <slot>"
 	local slot=${1}
 
-	local llvm_path=${ESYSROOT}/usr/lib/llvm/${slot}/bin
+	local llvm_path=${prefix}/usr/lib/llvm/${slot}/bin
```
[the remark this answers is not in the document]
Comment (mgorny) [reply eclass/llvm-utils.eclass:147]: It's for backwards compatibility with existing ebuilds. The new eclass doesn't use the `-d` version.

File: eclass/llvm-r2.eclass (hunk 2 above)
Comment (mgorny) [inline eclass/llvm-r2.eclass:390]: As pointed out in #39749, this affects CC/CXX lookup as well, so setting it is wrong. My next best idea is setting the bunch of `LLVM_ROOT`, `Clang_ROOT` (and so on) variables for `find_package()`.

**Comment by mgorny** [comment]
Fixed reported issues and updated to use `LLVM_ROOT` ,`Clang_ROOT` and `LLD_ROOT` vars.

**Review by chewi — APPROVED** [review]

File: eclass/llvm-r2.eclass [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+	debug-print-function ${FUNCNAME} "$@"
+
+	local esysroot_prefix=$(get_llvm_prefix -d)
+	einfo "Using ${esysroot_prefix} for CHOST LLVM ${LLVM_SLOT}"
+	[[ -d ${esysroot_prefix} ]] ||
+		die "LLVM ${LLVM_SLOT} not found installed in ESYSROOT (expected: ${esysroot_prefix})"
+
+	# satisfies find_package() in CMake
+	export LLVM_ROOT="${esysroot_prefix}"
+	export Clang_ROOT="${esysroot_prefix}"
```
Comment (thesamesam) [inline eclass/llvm-r2.eclass:427]: At e.g. https://codesearch.debian.net/search?q=CLANG_ROOT&page=4, I see some use of `CLANG_ROOT` too.

File: eclass/llvm-r2.eclass [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+	if [[ ! -f ${cmake_conf} ]]; then
+		cat <<-EOF
+			#!/usr/bin/env sh
+			echo "LLVM ${LLVM_SLOT} not installed for ABI=${ABI}" >&2
+			exit 127
+		EOF
+		return
+	fi
+
+	local version=$(
```
Comment (thesamesam) [inline eclass/llvm-r2.eclass:249]: Might be a bit neater with an array of vars and "$!X" check but not important if you disagree

File: eclass/tests/llvm-r2.sh [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+test_gen_dep() {
+	local arg=${1}
+	local expected
+	read -r -d '' expected
+
+	tbegin "llvm_gen_dep ${arg}"
+	local value=$(llvm_gen_dep "${arg}")
+
+	if [[ ${value} != ${expected} ]]; then
+		eerror "python_get_usedep ${arg}"
```
Comment (thesamesam) [inline eclass/tests/llvm-r2.sh:57]: python?

File: eclass/llvm-r2.eclass (hunk 4 above)
Comment (mgorny) [reply eclass/llvm-r2.eclass:427]: https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure

`<PackageName>_ROOT` takes precedence over `<PACKAGENAME>_ROOT`.

File: eclass/llvm-r2.eclass (hunk 5 above)
Comment (mgorny) [reply eclass/llvm-r2.eclass:249]: Hmm, is there a check that would check if all elements are non-empty?

File: eclass/tests/llvm-r2.sh (hunk 6 above)
Comment (mgorny) [reply eclass/tests/llvm-r2.sh:57]: Same in llvm-r1, will fix.

**Pull request closed without merging**